File: Design-Patterns-Using-Swift/SOLID/OpenClosedPrinciple.swift
class Animal {
    var name: String

    init(name: String) {
        self.name = name
    }
}

class Bird: Animal {
    override init(name: String = "") {
        super.init(name: name)
    }
        
    func fly() {
        // Abstract method, to be overridden by subclasses
    }
}

class Dodo: Bird {
    override init(name: String = "") {
        super.init(name: "Dodo")
    }

    override func fly() {
        print("The dodo is extinct and cannot fly.")
    }
}

class Penguin: Bird {
    override init(name: String = "") {
        super.init(name: "Penguin")
    }

    override func fly() {
        print("The penguin cannot fly.")
    }

    func slide() {
        print("The penguin is sliding on its belly!")
    }

    func swim() {
        print("The penguin is swimming in the water!")
    }
}

class Eagle: Bird {
    override init(name: String = "") {
        super.init(name: "Eagle")
    }

    override func fly() {
        print("The eagle is soaring through the sky!")
    }
}

class Sparrow: Bird {
    override init(name: String = "") {
        super.init(name: "Sparrow")
    }

    override func fly() {
        print("The sparrow is fluttering its wings!")
    }
}

class Pigeon: Bird {
    override init(name: String = "") {
        super.init(name: "Pigeon")
    }

    func makeCooingSound() {
        print("The pigeon is making a cooing sound.")
    }

    override func fly() {
        print("The pigeon is fluttering its wings!")
    }
}

// Client code
let bird1: Bird = Eagle()
bird1.fly()

let bird2: Bird = Dodo()
bird2.fly()

let bird3: Bird = Pigeon()
bird3.fly()

/*
The eagle is soaring through the sky!
The dodo is extinct and cannot fly.
The pigeon is fluttering its wings!
*/

File: Design-Patterns-Using-Ruby/SOLID/OpenClosedPrinciple.rb
class Animal
    attr_reader :name

    def initialize(name)
        @name = name
    end
end

class Bird < Animal
    def fly
        # Abstract method, to be overridden by subclasses
    end
end

class Dodo < Bird
    def initialize
        super("Dodo")
    end

    def fly
        puts "The dodo is extinct and cannot fly."
    end
end

class Penguin < Bird
    def initialize
        super("Penguin")
    end

    def fly
        puts "The penguin cannot fly."
    end

    def slide
        puts "The penguin is sliding on its belly!"
    end

    def swim
        puts "The penguin is swimming in the water!"
    end
end

class Eagle < Bird
    def initialize
        super("Eagle")
    end

    def fly
        puts "The eagle is soaring through the sky!"
    end
end

class Sparrow < Bird
    def initialize
        super("Sparrow")
    end

    def fly
        puts "The sparrow is fluttering its wings!"
    end
end

class Pigeon < Bird
    def initialize
        super("Pigeon")
    end

    def make_cooing_sound
        puts "The pigeon is making a cooing sound."
    end

    def fly
        puts "The pigeon is fluttering its wings!"
    end
end

# Client code
bird1 = Eagle.new
bird1.fly

bird2 = Dodo.new
bird2.fly

bird3 = Pigeon.new
bird3.fly

=begin 
The eagle is soaring through the sky!
The dodo is extinct and cannot fly.
The pigeon is fluttering its wings!
=end